Add Text Template guide to common widget properties

weirdwater opened the pull request:
Proposed changes for the Client Template placeholders. 

- [x] Copy and rewrite existing explanation from Text Widget
- [x] Add references to the Text Template section where templates are used.
- [x] Remove old explanation from Text Widget and add reference to Text Template section.

weirdwater committed:
Add Text Template guide to common widget properties

weirdwater commented:
Requested technical review from team.

weirdwater committed:
Update template explanations to refer to Text Template section

mlaponder commented on content/en/docs/refguide/modeling/pages/common-widget-properties.md:
```diff
@@ -403,3 +397,83 @@ The widget can be made visible to a specific of the user roles available in your
+
+## 10 Common Property Types
+
+This section details common types of inputs that share common behavior across different page properties.
```
"page properties" might be confusing (as properties of a page is a concept in SP), maybe  "widget properties" is better

weirdwater replied:
I used page properties because I wanted to stay generic to the Page Editor (as it is the subject of the page) and widgets are also configured using the page editor. It is also conceivable to me that the Text template could be used to configure a page title in the future.

Perhaps an even more generic version could be more fitting:

> This section details common inputs that share behavior, found in the page editor.

mlaponder replied:
✨

mlaponder commented on content/en/docs/refguide/modeling/pages/common-widget-properties.md:
```diff
@@ -403,3 +397,83 @@ The widget can be made visible to a specific of the user roles available in your
+
+### 10.1 Text Template {#text-template}
+
+Properties using the Text template offer the ability to display dynamic text on pages. This is done by writing templates with placeholders and specifying parameters. During runtime the placeholders are substituted by the parameters. The template editor consists of a minimized view and a pop-up view.
```
What are you referring to here with "minimized view"? The properties pane? (In that case, that is such a common concept I am not sure it is worth a specific mention here)

weirdwater replied:
I was indeed referring to the properties pane. I removed this section and updated a later reference.

mlaponder replied:
✨

mlaponder commented on content/en/docs/refguide/modeling/pages/common-widget-properties.md:
```diff
@@ -403,3 +397,83 @@ The widget can be made visible to a specific of the user roles available in your
+
+##### 10.1.3.1 Adding New Parameters
+
+To be able to use parameters, the widget must be in the context of an entity -- such as a [Data Widget](/refguide/data-view). Then to use them:
```
Parameters can also be used outside the context of an entity by referring to a page or snippet parameter which would be a nice addition here.

weirdwater replied:
Nice addition, I added them to the examples.

mlaponder replied:
✨

weirdwater committed:
Process feedback

weirdwater committed:
Make the section description more abstract

weirdwater commented:
Hi @ConnorLand, this pull request is now ready for your review.

ancajuravle commented:
And this comes with Mendix SP version 10.11.0 (to be released).

ConnorLand committed:
Format and edit

ConnorLand commented on content/en/docs/refguide/modeling/pages/button-widgets/button-properties.md:
```diff
@@ -62,52 +62,9 @@ The **Role type** property determines the `aria-role` attribute value that will
 #### 2.5.1 Caption {#caption}
 
-The **Caption** property defines a text that will be shown on the button. The caption can contain parameters that are written between braces, for example, {1}.  
+The **Caption** property defines a text that will be shown on the button using a Text template.
```
I've consistently bolded properties, but I'll check to see if that's correct.

ConnorLand committed:
Edit placeholder link text

ConnorLand committed:
Add bold formatting to UI buttons

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/button-widgets/button-properties.md:
```diff
@@ -62,52 +62,9 @@ The **Role type** property determines the `aria-role` attribute value that will
 
-For more information on using parameters, see the [Parameters](#parameters) section below. 
-
-#### 2.5.2 Parameters {#parameters}
```
If we are removing parameters from the Button properties, we need to add cross-references to the new place.

weirdwater replied:
This should be covered by the [inlined snippet](https://github.com/mendix/docs/pull/7745/files#diff-0a81bac9a4c697ab34be4e099387fc42115608c7e8829f1f7117c0d7740c087dR67).

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/common-widget-properties.md:
```diff
@@ -403,3 +397,83 @@ The widget can be made visible to a specific of the user roles available in your
 | All roles         | The widget is always visible. |
 | Selected roles    | This setting will render the widget as invisible to all users that are not linked to one of the selected user roles. |
+
+## 10 Common Property Types
```
Not sure I understand the name of this section and its explanation. The rest of the document is divided per section as they appears in **Properties**, this section is about parameters that are also in **Properties**.

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/common-widget-properties.md:
```diff
@@ -403,3 +397,83 @@ The widget can be made visible to a specific of the user roles available in your
+
+## 10 Common Property Types
+
+This section explains common inputs that share behavior, found in the page editor.
```
If we add this to the Common Properties doc, I would add a list of widgets that use parameters and would add cross-references in the corresponding documents of these widgets.

MariaShaposhnikova commented on static/_includes/refguide/text-template-link.md:
```diff
@@ -0,0 +1,3 @@
+
+For more information on Text templates, see the [Text Template Section](/refguide/common-widget-properties/#text-template) section in *Properties Common in the Page Editor*. 
```
Why do we call it "Text templates"? I did not see in the UI that the dialog box is called like this and I could not find a setting with this name. Will users understand that they need to look for Text templates to read about parameters?

MariaShaposhnikova replied:
```suggestion
For more information on text templates, see the [Text Template Section](/refguide/common-widget-properties/#text-template) section in *Properties Common in the Page Editor*. 
```

MariaShaposhnikova replied:
```suggestion
For more information on Text templates, see the [Text Template](/refguide/common-widget-properties/#text-template) section in *Properties Common in the Page Editor*. 
```

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/common-widget-properties.md:
```diff
@@ -403,3 +397,83 @@ The widget can be made visible to a specific of the user roles available in your
+
+This section explains common inputs that share behavior, found in the page editor.
+
+### 10.1 Text Template {#text-template}
```
Why do we call it "Text templates"? I did not see in the UI that the dialog box is called like this and I could not find a setting with this name. Will users understand that they need to look for Text templates to read about parameters?

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/common-widget-properties.md:
```diff
@@ -403,3 +397,83 @@ The widget can be made visible to a specific of the user roles available in your
+
+Properties using **Text Template** allows dynamic text to be displayed on pages. This is done by writing templates with placeholders and specifying parameters. During runtime, the placeholders are substituted by the parameters.
+
+#### 10.1.1 Template
```
For which widgets is Template used? I can see "Caption" using parameters as well.

MariaShaposhnikova reviewed (commented):
@ConnorLand , I've added a few comments on the PR. I tested this with a button and a text widget. Please test as well on these widgets and on other widgets that have parameters in their properties to unify the description that will fit all the widgets using parameters.

weirdwater commented:
Hi @MariaShaposhnikova, 

Thanks for the review and raising these questions. As some of the threads seem related I thought I would address them in a single comment.

**Why choose the term Text Template?**

The properties updated in this PR all use the same underlying property type, called [TextTemplate in the Pluggable Widgets Property Types](https://docs.mendix.com/apidocs-mxsdk/apidocs/pluggable-widgets-property-types/#texttemplate).

It is true that this is badly reflected in the UI. The **Fallback Text** help text in the dialog box does mention the term template, but the other fields do not. It would be more consistent and searchable if we updated the label and help text of the first field to refer to Template (see the underlined text in the screendshot).

![Screenshot 2024-06-05 at 13 49 03](https://github.com/mendix/docs/assets/8361804/0bbccce7-4381-4ac2-bf39-faedf690b4b6)

This change would also reflect the underlying datamodel more closely as a Placeholder (being edited in the screenshot) consists of a Template string, list of Parameters, and a Fallback text.

**Edit**: We have added a [story to our backlog](https://mendix.atlassian.net/browse/WTF-1896) to make the wording in the dialog box consistent.

**Why add the section Common Property Types?**

While researching the available documentation for the Text Template we noticed that it was spread across a number of pages. As the Text Widget's Caption section contained the most elaborate explanation, some sections in other files referred to it. Whereas the Label Caption and Button Caption sections had their own explanation.

As the type of this property is common to these properties it made sense to add a canonical description of the Text Template's usage. As it was describing a type of property rather than a specific property it did not seem to fit in an existing section.

The location was briefly discussed with Connor. The agreement was to place it in Properties Common to the Page Editor during writing and evaluate its location during the review.

**What widget properties use the Text Template property type?**

This is the list of properties using Text Template that I could find on short notice:

- Placholder property of:
  - [text-box](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/text-box.md)
  - [date-picker](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/date-picker.md)
  - [text-area](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/text-area.md)
- Label Caption property of:
  - [check-box](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/check-box.md)
  - [date-picker](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/date-picker.md)
  - [radio-buttons](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/radio-buttons.md)
  - [drop-down](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/drop-down.md)
  - [reference-selector](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/reference-selector.md)
  - [text-box](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/text-box.md)
  - [input-reference-set-selector](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/input-reference-set-selector.md)
  - [text-area](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/input-widgets/text-area.md)
  - [image-uploader](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/image-and-file-widgets/image-uploader.md)
  - [file-manager](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/image-and-file-widgets/file-manager.md)
  - [password-text-box](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/authentication-widgets/password-text-box.md)
  - [login-id-text-box](https://github.com/mendix/docs/tree/development/content/en/docs/refguide/modeling/pages/authentication-widgets/login-id-text-box.md)
  - Any Pluggable Widget with the [Label property type](https://docs.mendix.com/apidocs-mxsdk/apidocs/pluggable-widgets-property-types/#label).
- Caption property of Text Widget
- Caption property of Button Widget
- Data Grid 2 Column Properties:
  - [Dynamic Text](https://docs.mendix.com/appstore/modules/data-grid-2/#42-dynamic-text)
  - [Caption](https://docs.mendix.com/appstore/modules/data-grid-2-properties/#313-caption)
  - [Tooltip](https://docs.mendix.com/appstore/modules/data-grid-2-properties/#314-tooltip)
- Any Pluggable Widget with the [TextTemplate property type](https://docs.mendix.com/apidocs-mxsdk/apidocs/pluggable-widgets-property-types/#texttemplate).

weirdwater committed:
Refer to appropriate version in Mendix 8 and 9 documentation

ConnorLand commented:
@weirdwater thank you for the explanation. Doc changes are mostly targeting the right docs/locations, and I'll hone them a bit more per Maria's comments.

I think one of the major sticking points though is the fact that the property is not properly named (is now **Placeholder** instead of **Template** (maybe it would be even better to directly label it **Text Template** so docs and prop name are exactly synced?)). 

Would it be possible to finish [this story](https://mendix.atlassian.net/browse/WTF-1896) updating the label title before we merge docs?

weirdwater commented:
@ConnorLand I will add a link to the story to remember this PR when we pick it up. I hope to discuss it at our next refinement and have it in before 10.12.

MariaShaposhnikova commented:
@weirdwater , thanks for the detailed explanation! 
The PR says: **Caption** defines a text that will be shown using a Text template. And it also says: "The TextTemplate property allows a user to configure a translatable text template similar to the [Caption](/refguide8/text#caption) of a text widget."

As I understand, these are two different properties. Some widgets have a property named **Text template** and some **Caption**. Therefore, it is confusing for users when they read "*Caption** defines a text that will be shown using a Text template." The questions that I have: What is a text template? I don't see it in the UI in the properties, where do I find it?

The description of the **Common Properties** ideally should fit all the widgets that where we cross-reference it. So, the widgets that have a **Caption** property should cross-reference the **Caption** property description, and the ones that have **Text template** or a **Placeholder** should cross-reference those property descriptions. 
CC: @ConnorLand

weirdwater commented:
Hi @MariaShaposhnikova, I believe you are referring to the Pluggable Widgets Property Types for Mendix [8](https://github.com/mendix/docs/pull/7745/files#diff-b3e05aa00b94aede2d87bb8301e1760c30cc4c1bd79ce8a1930ba47ad26219e7) and [9](https://github.com/mendix/docs/pull/7745/files#diff-9d2cac71c178fbaa439f4a1e547b69780ad783314fc46c1a6316d514dcf3f6e6). They are part of the diff as they were referring to the Mendix 10 version of the text widget instead of their respective version. The equivalent documentation for [Mendix 10](https://github.com/mendix/docs/pull/7745/files#diff-a32d0cd057776acb0d6b03a8b80346e27e46b28a095f59ee493d74b45fccb138) mentions the Caption as a well known example but does not refer to it for reference.

It is not so much that a widget or page would have a property called Text Template. Instead they have a property (i.e. Caption, Placeholder, Dynamic Text) that is of the type Text Template. As a result each of the properties Caption, Placeholder, Dynamic Text, etc. has the sub-properties Template Text, Parameters, and Fallback Text. Currently the captioning in the dialog box does not reflect this well. We will decide on appropriate changes to the interface or documentation in an upcoming refinement of [WTF-1896](https://mendix.atlassian.net/browse/WTF-1896).

weirdwater commented:
Hi @ConnorLand and @MariaShaposhnikova, we have merged our changes to the UI for 10.12. The first field of the Text Template dialog is now labeled **Template** and is consistent across all usages of the property type. The user can still see what widget property they are editing in the title bar of the dialog.

![Screenshot 2024-06-14 at 15 58 15](https://github.com/mendix/docs/assets/8361804/fbe3be83-5f2b-44fd-9881-67f9d8c45203)

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/common-widget-properties.md:
```diff
@@ -403,3 +397,83 @@ The widget can be made visible to a specific of the user roles available in your
+
+{{< figure src="/attachments/refguide/modeling/pages/text-widgets/text/caption-edit-button.png" alt="Opening Parameters" width="450" >}} 
+
+#### 10.1.3 Parameters {#text-template-parameters}
```
```suggestion
#### 10.1.2 Parameters {#text-template-parameters}
```

MariaShaposhnikova replied:
I suggest we cross-reference this information from a separate snippet for the **Parameters** property.

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/text-widgets/text.md:
```diff
@@ -39,60 +39,10 @@ Styling:
 #### 3.1.1 Caption {#caption}
 
-**Caption** defines a text that will be shown. The caption can contain parameters that are written between braces, for example, {1}.  
+**Caption** defines a text that will be shown using a Text template. 
```
```suggestion
**Caption** defines a text that will be shown using a text template. 
```

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/common-widget-properties.md:
```diff
@@ -209,15 +209,9 @@ This property determines whether the label is rendered and the widget is wrapped
 ### 6.2 Label Caption
 
-This property is shown only when Show label is Yes. It specifies what text is rendered within a label.
+This property is shown only when Show label is Yes. It specifies what text is rendered within a label using a Text template.
```
```suggestion
This property is shown only when **Show label** is **Yes**. It specifies what text is rendered within a label using a text template.
```

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/text-widgets/text.md:
```diff
@@ -39,60 +39,10 @@ Styling:
+**Caption** defines a text that will be shown using a Text template. 
 
-For more information on using parameters, see the [Parameters](#parameters) section below. 
+{{% snippet file="/static/_includes/refguide/text-template-link.md" %}}
```
```suggestion
{{% snippet file="/static/_includes/refguide/text-template-link.md" %}}

#### 3.1.2 Parameters {#parameters}

For more information on parameters, see the [Parameters](/refguide/common-widget-properties/#text-template-parameters) section in *Properties Common in the Page Editor*.
```

MariaShaposhnikova commented on content/en/docs/refguide/modeling/pages/button-widgets/button-properties.md:
```diff
@@ -62,52 +62,9 @@ The **Role type** property determines the `aria-role` attribute value that will
-* **Move down** – to move a parameter down in the list of parameters and also to change its index, click **Move down**
-
-    {{< figure src="/attachments/refguide/modeling/pages/button-widgets/button-properties/button-parameter-actions.png" alt="Parameter Actions" class="no-border" >}}
+{{% snippet file="/static/_includes/refguide/text-template-link.md" %}}
```
```suggestion
{{% snippet file="/static/_includes/refguide/text-template-link.md" %}}

#### 2.5.2 Parameters {#parameters}

For more information on parameters, see the [Parameters](/refguide/common-widget-properties/#text-template-parameters) section in *Properties Common in the Page Editor*.
```
```

MariaShaposhnikova commented on static/_includes/refguide/placeholder-section-link.md:
```diff
@@ -0,0 +1,7 @@
+
+It can be used, for example, to give a hint to the end-user what kind of text should be entered.
+
+The Placeholder is rendered using a Text template. For more information see the [Text Template Section](/refguide/common-widget-properties/#text-template) section in *Properties Common in the Page Editor*.
```
```suggestion
The Placeholder is rendered using a text template. For more information see the [Text Template](/refguide/common-widget-properties/#text-template) section in *Properties Common in the Page Editor*.
```
